llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp

```diff
@@ -2150,13 +2150,42 @@ static std::optional<Instruction *> instCombineSVEInsr(InstCombiner &IC,
   return std::nullopt;
 }
 
+static std::optional<Instruction *> instCombineDMB(InstCombiner &IC,
+                                                   IntrinsicInst &II) {
+  // If this barrier is post-dominated by identical one we can remove it
+  auto *NI = II.getNextNonDebugInstruction();
+  int LookaheadThreshold = 10;
+  auto CanSkipOver = [](Instruction *I) {
+    return !I->mayReadOrWriteMemory() && !I->mayHaveSideEffects();
```

Comment:
Perhaps I'm being paranoid but what about things like `willreturn` because the currently implementation means a `dmb` before a non-returning call will never get executed.  This might signify user error, but given a missing `dmb` will be very hard to track down I feel like it's worth playing it say.

Does your use case concern crossing (non intrinsic) function call boundaries, if not then probably best to consider them as something that cannot be skipped over?

Reply:
`willreturn` should be handled by side effects check, IIUC:
```cpp
bool Instruction::mayHaveSideEffects() const {
  return mayWriteToMemory() || mayThrow() || !willReturn();
}
```